src/test/java/ArrayMinHeapTest.java

```diff
@@ -135,6 +157,25 @@ void peek_many_unalteredHeap() {
                             .usingRecursiveComparison()
                             .isEqualTo(preState);
                 }
+
+                @Test
+                void remove_many_returnsElementsCorrectOrder() {
+                    SoftAssertions softie = new SoftAssertions();
+                    softie.assertThat(classUnderTest.remove()).isEqualTo(5);
+                    softie.assertThat(classUnderTest.remove()).isEqualTo(9);
+                    softie.assertThat(classUnderTest.remove()).isEqualTo(10);
+                    softie.assertThat(classUnderTest.remove()).isEqualTo(10);
+                    softie.assertThat(classUnderTest.remove()).isEqualTo(15);
+                    softie.assertAll();
+                }
+
+                @Test
+                void remove_many_alterHeap() {
+                    classUnderTest.remove();
+                    assertThat(classUnderTest)
+                            .usingRecursiveComparison()
+                            .isNotEqualTo(preState);
+                }
```

Comment:
I suppose a slightly better version would be to check that the size is 1 less to check that _one_ thing was removed. It's still not a particularly effective test though.

Reply:
Test coupled methods in a workflow, not in a unit test.

Reply:
Alternitively, you could define a custom equals for heaps, scoped to your tests. Heaps would be equal to each other if the order of items returned by `remove` is equivalent. So, you could `remove` all elements from both, put those into lists, and assert equality on the lists. That is complex enough to warrant it's own tests though, so I'd prefer to see some workflow style tests instead.

Reply:
I am not familiar with what a workflow style test _means_. 

Do you mean something like 

```java
@Test
void remove_allElements_emptyHeap() {
    classUnderTest.remove();
    classUnderTest.remove();
    classUnderTest.remove();
    classUnderTest.remove();
    classUnderTest.remove();
    assertThat(classUnderTest).usingRecursiveComparison().isEqualTo(new ArrayMinHeap<>());
}
```

Reply:
Collections are hard to test. Methods like `remove` has some defined functional behaviour (which you can unit test), but they also have an _effect_ (which you can't **unit** test). By workflow, I mean invoking a series of methods, illustrating how methods are coupled, making several assertions along the way. It's a literal integration test, but people (at work) conflate that term with service tests all the time, so I've trained myself to avoid using the term.

For example, `add` has a functional behaviour, it returns `true`. It also has effects, which are usually _`someMethod` maybe returns something different_. `someMethod` then has some effects... and so on. You'd capture these in integration tests. Something like the following, given an empty heap for the `classUnderTest`:
```java
@Test
void add_onlyUseCase_effectsContents() {
    classUnderTest.add(10);
    // this list of assertions is illustrative; you'd assert for every coupled method
    assertThat(classUnderTest).hasSize(1);
    assertThat(classUnderTest).contains(10);
    assertThat(classUnderTest.remove()).isEqualTo(10); // careful, remove has effects!
}
```

Note that there are a lot of cases here. When `classUnderTest` is a singleton, you can add to the front or to the back, changing the effect on the `remove` result. You can also add a duplicate to a singleton. When it's larger, there's more cases still (add to the middle, duplicate front, duplicate back, duplicate middle). And, that is only `add`. You have many similar effects for `remove`. It's a shitload of work.

Stateful classes, am I right?

We can be lazy, though. Good enough is good enough. We can test a lot of cases at once, especially in the large heap case (just add many items). In the example above, we called `remove` in an assertion, we could continue from there, too.

```java
@Test
void integration_onlyUseCase_effectsContents() {
    classUnderTest.add(10);
    assertThat(classUnderTest).hasSize(1);
    assertThat(classUnderTest).contains(10);

    // removal of the only element just added
    assertThat(classUnderTest.remove()).isEqualTo(10);

    assertThat(classUnderTest).hasSize(0);
    assertThat(classUnderTest).doesNotContain(10);
}
```

You'll note that, now, the second half of that test is no longer testing an empty heap, so you will lose organization you gain from `@Nested`. It's up to you how you want to actually organize everything. I do tend to see separate integration test classes (so `ArrayMinHeapIntegrationTest`). (You can keep it in the same file though, thanks junit5!)